List View and Block List: Update hover, focus, highlight, select borders

Copons opened the pull request:
## Description

Fixes #29467

The initial implementation of the Persistent List View had a pretty big usability issue caused by the fact that the component wasn't designed to stick around after an item was selected.
The persistence required a visual clue to indicate that, on item selection, the focus shifted from the list to the block.

This PR subtly updates the outlines of both the List View items, and the blocks in the canvas, to make their relationship clearer.

- In canvas, the primary selection (selected, hovered, highlighted) outline changed from `2px --wp-admin-theme-color` (typically blue) to `1px $gray-900`.
- In canvas, the Template Part block's outline becomes `dotted` when one of its children is selected.
- In the List View, focused items gain a white halo inside their blue outline.

This PR also highlight the entire children branch of a selected item.
This required the introduction of some additional logic in the `BlockNavigation` component.
To avoid regressions and incompatibilities, like the other Persistent List View-only features, this too is gated behind the `__experimentalPersistentListViewFeatures` flag.

## How has this been tested?

- Activate an FSE theme such as TT1 Blocks.
- Enter the Site Editor, and open the List View.
- Play around with it, making sure the focus position is clear enough at all times.

## Screenshots <!-- if applicable -->

Grey outline of a hovered block:

<img width="897" alt="Screenshot 2021-03-15 at 16 58 39" src="https://user-images.githubusercontent.com/2070010/111191308-d7218000-85af-11eb-997a-681a81b4d6ca.png">

Dotted outline of a template part with a child selected:
<img width="899" alt="Screenshot 2021-03-15 at 16 58 19" src="https://user-images.githubusercontent.com/2070010/111191341-df79bb00-85af-11eb-9d27-f40173fb512d.png">

List View item's focus halo:

<img width="332" alt="Screenshot 2021-03-15 at 16 59 13" src="https://user-images.githubusercontent.com/2070010/111191501-0a640f00-85b0-11eb-8477-f83e42e39714.png">

List View's selected branch:

<img width="329" alt="Screenshot 2021-03-15 at 16 57 51" src="https://user-images.githubusercontent.com/2070010/111191546-16e86780-85b0-11eb-88c8-39ca2b969a4f.png">

## Types of changes
New feature (non-breaking change which adds functionality)

## Checklist:
- [x] My code is tested.
- [x] My code follows the WordPress code style. <!-- Check code: `npm run lint`, Guidelines: https://developer.wordpress.org/coding-standards/wordpress-coding-standards/javascript/ -->
- [x] My code follows the accessibility standards. <!-- Guidelines: https://developer.wordpress.org/coding-standards/wordpress-coding-standards/accessibility/ -->
- [x] I've tested my changes with keyboard and screen readers. <!-- Instructions: https://github.com/WordPress/gutenberg/blob/HEAD/docs/contributors/accessibility-testing.md -->
- [x] My code has proper inline documentation. <!-- Guidelines: https://developer.wordpress.org/coding-standards/inline-documentation-standards/javascript/ -->
- [x] I've included developer documentation if appropriate. <!-- Handbook: https://developer.wordpress.org/block-editor/ -->
- [x] I've updated all React Native files affected by any refactorings/renamings in this PR. <!-- React Native mobile Gutenberg guidelines: https://github.com/WordPress/gutenberg/blob/HEAD/docs/contributors/native-mobile.md -->

Copons committed:
Update Persistent List View colors

Copons committed:
Update Block List colors

Copons committed:
Adjust focused+selected block

jameskoster commented:
This is awesome to see :)

Before getting in the _style_ of everything (which will certainly need some extra thought and refinement) I noticed one small issue:

If you select a block on the canvas then open list view, focus transfers to the list view as expected 👍 If you then close list view using the keyboard shortcut focus seems to vanish. Should it return to the selected block on the canvas? Video:

https://user-images.githubusercontent.com/846565/110361655-39223880-8038-11eb-91dd-42a8cdde9b61.mp4

jameskoster commented:
Another small issue that I _suspect_ is related – if you select a text-based block from list view, focus transfers to the canvas as expected, but for some reason I do not see the text cursor, and if I start typing characters appear at the _beginning_ of the value:


https://user-images.githubusercontent.com/846565/110362382-1f352580-8039-11eb-9f98-38796628f8b1.mp4

Copons commented:
> Before getting in to the _style_ of everything (which will certainly need some extra thought and refinement)

Yeah, there are a couple of minor things in the current code, but most importantly, I literally typed random block list classnames until they worked well enough.
There are _a lot_ of combinations, and many of them are obscure to me: navigation mode, edit mode, outline mode, focus mode, is selected, is highlighted, is hovered, :focus, is typing, is editing — and I'm pretty sure I'm missing more. 😅
I was mostly interested in having something to play around with.

> If you select a block on the canvas then open list view, focus transfers to the list view as expected 👍 If you then close list view using the keyboard shortcut focus seems to vanish. Should it return to the selected block on the canvas? Video:

Blurgh.
Keyboard navigation is too hard, we should just skip ahead to brain interfaces.

> Another small issue that I _suspect_ is related – if you select a text-based block from list view, focus transfers to the canvas as expected, but for some reason I do not see the text cursor, and if I start typing characters appear at the _beginning_ of the value:

I think the cursor at the beginning is unrelated to the Persistent List View, and it's just how the `selectBlock` action works (which is not to say it's correct 🤔 ). The normal List View does the same, for example.

The missing cursor might be hidden under one of the three outlines of the selected/focused block.
Although, the block shouldn't have that outline when `.is-typing`, but it's likely conflicting with other rules.

jameskoster committed:
Reduce focus width

paaljoachim commented:
Here is an overview:

https://user-images.githubusercontent.com/5323259/110508679-8cac8900-8101-11eb-80c6-a1774fa8e55a.mp4


I really like what I see. It gives a nice overview.

Copons commented:
@jameskoster I... I really can't see the difference with d772416, but I'm happy it saves a few characters! 😅 

Anyway, I think the changes introduced in this PR make things 100% clearer, at least in relation to the Persistent List View.

We'd be pushing major visual changes to the block editor itself though, so I'd expect more discussions from designers before committing to an approach.

Otherwise, I could limit the outline changes to only the Persistent version of the List View and only the Site Editor's canvas (so, basically, limiting all the changes to the Site Editor).
It would allow more people to check it out IRL and see how it feels.
Should we get some backlash, we can just simply revert.

jameskoster commented:
Exploring ways to eliminate the double-border on the canvas which looks kind of ugly. Perhaps when a block is selected + focussed we only show the focus treatment? I believe this still satisfies the accessibility requirements as the border width and position both change.


https://user-images.githubusercontent.com/846565/110661335-5f70e100-81bc-11eb-8f7e-815b37eca37b.mp4

jameskoster commented:
Having got some feedback from the design team, I think what we're doing in this PR may be overkill :) 

I think the chief concern for now should be addressing the issue in the following screenshot:

<img width="1422" alt="Screenshot 2021-03-10 at 19 53 38" src="https://user-images.githubusercontent.com/846565/110775649-10c45500-8257-11eb-8f83-eff4f71cf717.png">

Focus is in list view, but since blocks on the canvas with the `isHighlighted` (and `isHovered` for that matter) state receive the same style treatment, that is not clear and obvious.

The remedy for this seems fairly straight-forward – reserve the blue treatment for focus, and use another color (likely $gray-900) for hover and highlight states. The only drawback I can see to this is that template parts will then receive the same treatment (when a child is selected), as when they are hovered. So we may need to adjust that style as well.

I'll push some changes for us to test soon.

jameskoster committed:
Style adjustments

Hovered and highlighted blocks on the canvas now received a 1px dark grey outline rather than a 1.5px blue one. Blue is reserved for focus.

When a child is selected, the associated template part now receives a dotted border to differentiate from its hover and highlighted states.

jameskoster commented:
Pushed the changes suggested in my last comment. Here's a video demo:

https://user-images.githubusercontent.com/846565/110782516-250c5000-825f-11eb-863d-b49546d13d02.mp4

I suspect the dotted outline on the template part will be the most controversial change. There are perhaps other options to explore there, light setting an opacity instead of the different border-style.

jameskoster committed:
Remove template part border when highlighted

jasmussen commented:
I haven't fully tested this PR, as I'm in the middle of a separate one, but thank for all the work!

> I suspect the dotted outline on the template part will be the most controversial change. 

I'm generally skeptical of dotted and dashed lines, as they tend to add noise. However with borders and boundaries and hover states in the site editor still in need of a fair bit of holistic thinking and love, if we are to try a dotted line, this is as good a time as any. If it doesn't work, we can explore a semi-transparent border instead.

Copons commented:
@jameskoster your latest changes help a lot with the list/canvas focus confusion!
I'm still slightly tricked by the black background of selected list items, but at least seeing the blue border moving from list to canvas makes things pretty clear to me.

I don't have a strong opinion about the dotted template part border.
Dots and dashes have a bit of a nostalgic web 1.0 flavour, but I like that there's a distinction (even if subtle) between template parts and other blocks.

---

I've noticed two unrelated issues that become more obvious with the Persistent List View.
(I'll open issues for these shortly)

- The Post Title block's outline is bugged, and it's rendered as a 0x0 shape, which can be seen as a little dot in the top-left corner of the block (the dot should be the outset outline I think).

- The Query block plays a little wonky with the Persistent List View.
In the list, it only shows the nested blocks for the most recent post in the loop, whereas in the canvas there might be several posts.
Selecting one of the "other" posts updates the list view: the Query will show the nested blocks for the selected post in the loop.
I'd probably expect to see the whole loop "unfurled" in the list view. 🤔

jameskoster reviewed (approved):
I've noticed those issues as well. 

I think it is worth merging this. There will no doubt be subsequent adjustments to further refine these interactions, but this feels like a step forwards overall.

Copons commented:
Let me clean up the code a little before merging 🙂

Copons committed:
Code cleanup

Copons commented:
Just FYI in 3f7b4eb I've:

- Restored the selected branch background color (light gray).
- Changed the list item hover from a light gray background to a dark gray outline. (Note: this also affects non-persistent List Views)
- Stop marking the list item focus' double border as "experimental". (Note: this also affects non-persistent List Views)
- Removed an inline comment that's now incorrect.

Later I'll update the PR description and proceed with the merge. :)

jasmussen commented:
I don't think this one is ready. 

Here's before:

![before site editor](https://user-images.githubusercontent.com/1204802/110826551-7a635400-8295-11eb-9c55-d83dddeca28a.gif)

![before post editor](https://user-images.githubusercontent.com/1204802/110826557-7c2d1780-8295-11eb-871f-572db7c762ed.gif)

Here's after:

![after post editor](https://user-images.githubusercontent.com/1204802/110826564-7e8f7180-8295-11eb-9cff-4e34114426cd.gif)

![after site editor editor](https://user-images.githubusercontent.com/1204802/110826570-7fc09e80-8295-11eb-88a8-b19dcddb4009.gif)

- The hover was better when gray, the bordered hover is not used anywhere else in the editor
- The selected state halo inside the blue box shadow is not necessary, the focus style on a selected item is already a shape change
- I'm not convinced that showing the child blocks with a gray background actually helps legibility, and it's not something we use elsewhere.

Thank you for all the work, but I would not make any changes to how hover and focus works inside the list view, I'd keep the changes solely to inside the canvas.

jameskoster commented:
Two more very small details to address @Copons (sorry).

* The last item in a branch should have a rounded bottom.
* When an item _inside a branch_ is selected but has no children of its own, it should be fully rounded.

Copons commented:
@jameskoster @jasmussen If it's all right with you, I think it's probably easier to move the Block List changes into a different PR, merge that one, then rebase this and keep discussing/playing around with the List View style.

I'm totally cool with opening it first thing tomorrow, or @jameskoster could do it if you want to keep full authorship of your changes. 🙂

jameskoster commented:
@copons I think we can continue, this is close :D 

> The hover was better when gray, the bordered hover is not used anywhere else in the editor

I'll revert this and we can potentially revisit in the future.

> The selected state halo inside the blue box shadow is not necessary, the focus style on a selected item is already a shape change

We spoke about this on Slack and found some consensus based on the fact that the white halo effect is used on a few other situations like selected Icon Buttons:

<img width="109" alt="Screenshot 2021-03-11 at 17 44 19" src="https://user-images.githubusercontent.com/846565/110834205-2bb5ba00-8295-11eb-9a16-8f4abc71d0a7.png">

> I'm not convinced that showing the child blocks with a gray background actually helps legibility, and it's not something we use elsewhere.

I think this is particularly useful in this list view. In heavily nested blocks, when you scroll so that the parent is hidden (and indentation consequently becomes less intuitive), it helps to understand where the container ends. We could perhaps restrict this to list view only? I believe this is the final sticking point.

jameskoster committed:
Revert list view hover state to a solid background

Also adjusts the background color on children of the selected parent

jasmussen commented:
> I think we can continue, this is close :D

I think so too. I think the hover change is the one that needs further thought, as the additional borders feel like they detract from focus:

<img width="416" alt="Screenshot 2021-03-11 at 18 46 31" src="https://user-images.githubusercontent.com/1204802/110835404-08900800-829f-11eb-8c8f-3f21e94bb80f.png">

Changing that back to a gray background hover addresses the biggest point. The rest we can look at.

jameskoster commented:
Hover styling re-instated:

![hover](https://user-images.githubusercontent.com/846565/110835865-438e3d80-8297-11eb-8453-946043d168b0.gif)

jasmussen commented:
Cool. I'm still unsure about the background for children, and I think our focus styles could use a holistic looking at so we don't write 100 different focus style rules. But that seems easy to revisit.

Copons commented:
@jameskoster @jasmussen Just FYI the children background is only enabled for the Persistent List View, so it wouldn't affect pre-existing uses of `BlockNavigation`.

> * The last item in a branch should have a rounded bottom.
> * When an item _inside a branch_ is selected but has no children of its own, it should be fully rounded.

I'll see to take care of these!

Copons committed:
Fix border radius of selected terminal items

Addison-Stavlo committed:
className applied for last item in selected branch

Copons committed:
Code cleanup

Copons committed:
More clarifications

jeyip reviewed (commented):
## Testing

### Requirements
- [x] In canvas, the Template Part block's outline becomes dotted when one of its children is selected.
- [x] In the List View, focused items gain a white halo inside their blue outline.
- [ ] In canvas, the primary selection (selected, hovered, highlighted) outline changed to gray

@Copons I had a question about the canvas primary selection selected outline. I might be missing something, but could you elaborate on how to arrive in the selected state? I'm clicking on canvas blocks, and the outlines still appear to be blue?

### Browsers
- [x] Firefox
- [x] Edge
- [x] Safari
- [x] Chrome
- [ ] ~IE11~

**Note:** 
- Didn't test IE11 (see [here](https://github.com/WordPress/gutenberg/issues/28249)). It looks like a fix is currently in the works, so hopefully (or maybe not so hopefully? 😅)  we can test it again soon.
- There seems to be a noticeable delay when editing canvas text content with the persistent list view open. (Open the list view, click on a paragraph block, and start typing). Wondering if anyone else can reproduce this as well.
- I saw some general navigation block select state weirdness in Safari, but I could replicate it on the master branch as well. It appears unrelated to the changes made here. I'll look for an existing issue and file one if it doesn't already exist first thing tomorrow morning.

#### Selecting and unselecting nav links unexpectedly modifies link opacity
![2021-03-16 00 26 51](https://user-images.githubusercontent.com/5414230/111271997-3ce12c80-85ef-11eb-8e7d-f3be5c65641b.gif)

jeyip reviewed (commented):
The code itself looks reasonable to me 👍 

+1 from me when testing question is answered 🚀

Copons commented:
> @Copons I had a question about the canvas primary selection selected outline. I might be missing something, but could you elaborate on how to arrive in the selected state? I'm clicking on canvas blocks, and the outlines still appear to be blue?

@jeyip The blue outline is because the block is `:focus`ed.
It's super complicated to wrap the head around, but easy to test:
- Open the List View and click on an item.
- The corresponding block will be selected (grey outline) and focused (overriding blue outline).
- Click on some white space (e.g. the header toolbar, or the List View title rectangle) to draw the focus away from the block, while leaving it selected.
- The block is now just selected (grey outline).

Note that not all blocks work in the same way. Sometimes the blue outline show up when you start editing their text content. Sometimes they never have a blue outline somehow (the Navigation block, for example, has the blue outline, but the Navigation Link blocks inside it don't).

I'm not sure about the root of these inconsistencies, or if there's an intentional rule behind them.
Though, the PR doesn't really change their behaviours (as in, changing when a block is selected or focused), it only adjusts the states colors.

cc @jameskoster in case there's more to this. 🙂

Copons commented:
> There seems to be a noticeable delay when editing canvas text content with the persistent list view open. (Open the list view, click on a paragraph block, and start typing). Wondering if anyone else can reproduce this as well.

Seing this as well. 🤔 
It's unrelated to this PR though, so it's a bug of the Persistent List View itself.
Going to investigate. Thanks for reporting!

> I saw some general navigation block select state weirdness in Safari, but I could replicate it on the master branch as well. It appears unrelated to the changes made here. I'll look for an existing issue and file one if it doesn't already exist first thing tomorrow morning.

Yeah apparently there are several issues related to blocks states in Safari, as if a block state "sticks around" after the state changed.
Definitely happening in `trunk` as well, and anyway it's not something that would be affected by this PR.

jasmussen commented:
> Yeah apparently there are several issues related to blocks states in Safari, as if a block state "sticks around" after the state changed.

Is that related to #29677 and/or #29678 ?

Copons commented:
> > Yeah apparently there are several issues related to blocks states in Safari, as if a block state "sticks around" after the state changed.
> 
> Is that related to #29677 and/or #29678 ?

@jasmussen #29677!

jeyip commented:
> The corresponding block will be selected (grey outline) and focused (overriding blue outline).

Oh I see! There's a difference between select and focus, and the focus border styles will override the select border styles. Gotcha 👍

@jameskoster Out of curiosity, has there been a discussion about the difference between select and focus states in Gutenberg outlined anywhere? Is it similar to what [this external article ](https://docs.microsoft.com/en-us/windows/win32/winauto/selection-and-focus-properties-and-methods#:~:text=There%20are%20some%20key%20differences,system%20that%20receives%20keyboard%20input.&text=A%20selected%20object%20is%20marked,some%20type%20of%20group%20operation.) describes?

- A focused object is the one object in the entire operating system that receives keyboard input. The object with the keyboard focus is either the active window or a child object of the active window.
- A selected object is marked to participate in some type of group operation.

jameskoster commented:
@jeyip good question! Not that I am aware of. The descriptions you shared seem quite closely aligned with what we have in the block editor though. 

Clicking an object (an on-canvas block in this case) will select _and_ focus it. A selected and focussed block can be directly manipulated, IE when you click a paragraph block you can begin typing immediately.

The focus state can move independently of the select state to a variety of different elements such as the block toolbar, the list view, or the inspector. In all of these cases the on-canvas block remains selected (and we need to indicate this visually) so that users are able to comprehend that it is the target of operations contained within the aforementioned elements – so things like grouping, transforms, text formatting, layout adjustments, and so on.

I know y'all know this. But it can be easy to forget if you aren't performing these actions keyboard-only on a day to day basis 😅

vindl reviewed (approved):
This is a very nice improvement overall! 🚀 

I noticed two issues while testing, but both of them seem to be edge cases that can be addressed later in my opinion.


https://user-images.githubusercontent.com/1182160/111365998-46bc6b80-8693-11eb-83e3-84973c9a72cd.mp4

1. After opening the navigator with keyboard shortcuts and pressing tab a few times the footer TP is selected, but its contents are not rendered properly and gray background is shown instead.

2. After selecting the header for example and using down arrow to focus other top level groups, the focused item is not scrolled into view and goes outside of the viewable area.

Copons commented:
> 1. After opening the navigator with keyboard shortcuts and pressing tab a few times the footer TP is selected, but its contents are not rendered properly and gray background is shown instead.

@vindl The `TreeGrid` component (which powers `BlockNavigation`) is navigable with arrows instead of tab, so when you tab, you are actually moving out of the list and, in this case, into the canvas, selecting the top level blocks.
For some reason, the content below the fold doesn't render when navigating via tab.
I'd say it's not related to the List View at all, though; most likely it's a Site Editor issue. 🤔 

> 2. After selecting the header for example and using down arrow to focus other top level groups, the focused item is not scrolled into view and goes outside of the viewable area.

That's because list scroll and canvas scroll are not "in sync".
After you tab away from the List View and into the canvas, you end up navigating through the canvas.
Eventually you find your way to the Footer. The canvas scrolls appropriately (even though the content is not rendered); but the List doesn't scroll at all.

It's something that bugs me as well, to be honest, and I've raised it in the past.
We eventually decided to not do anything about it, as it might be counterproductive.
For example, you might want to compare the list at a scroll point with the canvas at another scroll point, or (in the future) drag a block from the list to the canvas.
Locked scrolls might get in the way, or even make it impossible altogether.

Copons commented:
This PR subtly updates the outlines of both the List View items, and the blocks in the canvas, to make their relationship clearer.

- In canvas, the primary selection (selected, hovered, highlighted) outline changed from `2px --wp-admin-theme-color` (typically blue) to `1px $gray-900`.
- In canvas, the Template Part block's outline becomes `dotted` when one of its children is selected.
- In the List View, focused items gain a white halo inside their blue outline.

This PR also highlight the entire children branch of a selected item.
This required the introduction of some additional logic in the `BlockNavigation` component.
To avoid regressions and incompatibilities, like the other Persistent List View-only features, this too is gated behind the `__experimentalPersistentListViewFeatures` flag.

vindl commented:
> @vindl The TreeGrid component (which powers BlockNavigation) is navigable with arrows instead of tab, so when you tab, you are actually moving out of the list and, in this case, into the canvas, selecting the top level blocks.

I tested the arrow navigation and tried this after - with the current state/styles it _appears_ as though it's working with tab too, even though under the hood it's focusing things in the canvas, so I got confused by it. :)

> For example, you might want to compare the list at a scroll point with the canvas at another scroll point, or (in the future) drag a block from the list to the canvas. Locked scrolls might get in the way, or even make it impossible altogether.

Thanks for clarifying, makes sense now.
